### DISCUSSION TITLE
Tips for model training

### DISCUSSION THREAD

**Discussion by SebTim** [neuropoly/axondeepseg#503] [Help needed] [answered]
Dear all, 

I am trying to train a new model on my own images. Despite using tensorflow-GPU it drops out most of the time after several epochs due to memory issues. I guess I am doing something wrong with the data itself or the config file. 
What general properties would you recommend to start from when training a new model?
I am wondering about optimal input image number, minimum magnification, image size, general pixel size, patch size, batch size and number of epochs. 

Another problem might be that I have images of cortical areas that are only sparsely myelinated. So many image patches might be "empty". Last time I used 24 images taken at 2.5K magnification (pixel size = 0.00596 um, 21.89 um x 21.89 um, 3673x3673 pix square) with corresponding masks. I tried different splittings in several hundred patches (e.g. 512x512 pix), but it usually drops out. I ran 600 epochs at a batch size of 8. I tried reducing the batch size, but that didnt help either.

It usually shows:

` W tensorflow/core/common_runtime/bfc_allocator.cc:267] Allocator (GPU_0_bfc) ran out of memory trying to allocate 128.00MiB.  ` 
followed by a list of chunks

the summary:

`
2021-04-09 10:07:57.318438: I tensorflow/core/common_runtime/bfc_allocator.cc:647] Stats:
Limit:                  9207214244
InUse:                  9051609600
MaxInUse:               9148100352
NumAllocs:                   34914
MaxAllocSize:           3355443200
`

and then:

` W tensorflow/core/framework/op_kernel.cc:1401] OP_REQUIRES failed at conv_grad_input_ops.cc:942 : Resource exhausted: OOM when allocating tensor with shape[8,16,512,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
`

and 

```
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[8,16,512,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
         [[{{node training/Adam/gradients/finalconv/convolution_grad/Conv2DBackpropInput}}]]

```

Do I have too many images, or is something wrong with the settings?

many thanks

Sebastian

**Comment by dyt811** [comment]
Sorry I am more or less out of context here as I have not played with ADS too much before. However, have you tried using much smaller tensor size instead? [2, 4, 128, 128]: start with a tiny batch size, and smaller image sizes as a start? When I was doing deep learning training before I typically find it easier to debug by starting with a smallest possible data input and train longer time to avoid OOM error and then start exploring scenarios where you can increase batch size or image resolutions. Also, usually bigger GPU RAM helps (even if slower, e.g. Titan Maxwell/Pascal with 12GB RAM while slower can typically allow you to throw more data at it at the same time  than 1080Ti 8GB for comparison. 

From purely base on the information provide (please take this with a grain of salt as I am missing much context), I suspect the image patch resolution is too high. Most computer vision deep learning application train on images in the 128px to MAYBE 256px size as raw input. 512 at a batch size of eight might really pushing RAM demand a lot so I would recommend if possible and applicable, try to start small maybe even at 64px just to test out the input output process before any meaningful training is further conducted. 

A really rough estimation of batch size can be conducted if you know the trainable parameters in the model used, as explained here: https://stackoverflow.com/a/46656508

**Comment by mariehbourget** [comment, marked as answer]
Hi Sebastian, thank you for reaching out!

With ADS, we typically train on image sizes of 256x256 or 512x512. We had better results in the past with 512x512 patches for our datasets. To help with the memory allocation issue, you could try with smaller patches of 128x128 but you have to make sure that you patches are big enough to have full axons inside of them.

To create the patch, you need to use the [01-guide_dataset_building.ipynb](https://github.com/neuropoly/axondeepseg/blob/master/notebooks/01-guide_dataset_building.ipynb) notebook.
The 2 parameters to set are:
`general_pixel_size` : resamples your images to a given resolution
`patch_size` : splits your images into workable patches

Depending on how large your images are and how big your axons are relative to your image, you might want to choose a `general pixel size` that downsamples your image, so that the model processes faster or most of your axons are not cut through by the patching step.

To adjust these parameters (for example, with a `patch_size` of 512x512), you need to use a `general_pixel_size` to obtain patches of at least 512 pixels. To do that, take the smallest dimensions of your image (here: 3673), multiply it by your `pixel size` (here: 0.00596 um) and divide it by the `general_pixel_size` (example: 0.01).  You would have `3673*0.00596/0.01 = 2189 > 512`.

With a higher `general pixel size`, you will have less patches and more axons per patch.
The idea is to use a `general_pixel_size` that would give you patches big enough to have full axons inside of them. We typically recommend about 4 to 20 patches per image but it can be more depending on the content. The example above gives 25 patches per image.

> Despite using tensorflow-GPU it drops out most of the time after several epochs due to memory issues.

I’ve never had this issue personally, when the batch size is too high, I have the `Resource exhausted: OOM when allocating tensor` error in the first epoch, but not after several epochs. How many epochs were you able to run with your 512x512 patches and batch_size of 8? Are you using the same network architecture as the one in the [02-training_guidelines.ipynb](https://github.com/neuropoly/axondeepseg/blob/master/notebooks/02-training_guideline.ipynb) notebook? 

As @dyt811 mentioned, I would recommend doing some tests with a small batch size (2-4) to have less data processed at the same time.
For the number of epochs, you could start with 300 instead of 600.

> Another problem might be that I have images of cortical areas that are only sparsely myelinated. So many image patches might be "empty".

This may not be an issue for GPU memory, however, we had a similar situation in the past and found we have better results with the training when removing those “empty” patches from the dataset.

> or is something wrong with the settings?

As a next step, can you send us an example image and mask, your config file, and the code used to create the patches and start the training so we can check the settings?

Also, are your images TEM? If so, have you tried our default_TEM_model?

Best,
Marie-Hélène

**Comment by Stoyan-I-A** [comment]
> it drops out most of the time after several epochs due to memory issues

I find it a bit weird that the program works for a few initial epochs but stops after a certain amount of them. 
If you open your task manager, on the performance tab, you can see how much memory your dedicated GPU is using. If the GPU memory used after many epochs is significantly higher than the GPU memory using during the first few epochs, there might be a memory leak somewhere.

**Reply by SebTimCam** [reply]
I think you are right. 
I now tried the example data and simply used all the default settings. I get the same problem that the training aborts. First after 18 epochs, after I restarted already after 3. The resourch monitor does not show high GPU, CPU or memory usage. 

Maybe this is the key error message?` 2021-05-11 22:21:47.502524: W tensorflow/core/framework/op_kernel.cc:1401] OP_REQUIRES failed at conv_grad_input_ops.cc:942 : Resource exhausted: OOM when allocating tensor with shape[8,16,512,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc`

**Reply by Stoyan-I-A** [reply]
Check specifically the GPU Memory with a small batch size. Check if the GPU Memory gets filled up too much over time.

**Comment by SebTimCam** [comment]
Dear dyt811, Marie-Helene and Stoyan, 

sorry for not getting back to you earlier. I am now back on the model training. 

> Also, are your images TEM? If so, have you tried our default_TEM_model?

Yes, I am using TEM images. I tried the default TEM model, it is working on some images. A little bit of preparing them beforehand (background substraction, filters etc.) helped to improve the results, but I thought I could get much better segmentation with my own model. So far it was not really successfull. 

I will now use the suggested parameters, monitor the memory and see where it gets me. I will keep you updated!

Sebastian

**Reply by mariehbourget** [reply]
Hi Sebastian,

Regarding the `default_TEM_model`, one of the things you can try to improve the segmentations is to try different pixel sizes when launching the segmentation.

In command line, you can do that by trying different values for the parameters SIZEPIXEL (`-s`),  see our doc here: https://axondeepseg.readthedocs.io/en/latest/documentation.html#syntax.

Alternatively, in the `00-getting_started.ipynb` notebook, you can achieve the same by adding the parameter `acquired_resolution` to the `axon_segmentation` function, where `acquired_resolution` is the pixel size you want to test. Example: 
```
prediction = axon_segmentation([path_testing], ["image.png"], path_model, config_network, acquired_resolution=0.005, resampled_resolutions=0.01, verbosity_level=3)
```
